src/libraries/System.Private.CoreLib/src/System/Utf8String.Comparison.cs

```diff
@@ -85,7 +85,7 @@ public static bool AreEquivalent(Utf8String? utf8Text, string? utf16Text)
         /// </remarks>
         public static bool AreEquivalent(ReadOnlySpan<byte> utf8Text, ReadOnlySpan<char> utf16Text)
         {
-            if (utf8Text.Length == 0 && utf16Text.Length == 0)
+            if (utf8Text.IsEmpty && utf16Text.IsEmpty)
             {
                 // Don't use IsEmpty for this check; JIT can optimize "Length == 0" better
                 // for this particular scenario.
```

Comment:
Apparently this case shouldn't change?  Which raises the question of "why" and whether that applies anywhere else (and how we can fix the JIT so that it's not a pit of failure either way).
cc: @GrabYourPitchforks

Reply:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunHjykDYBAgAbBgFkAFACUY2ACYB5AHaiAngGUADthUAeYOowwAfAwCuGAGYAOACowEGNAzmLVGnXv1gAFthQ5lbWAo7OAJRcPJzUPPEMAJbWDFIhDk4YAHQAMjAqAOYYfgwAvKUMAAwMAGQ1ljZhmbn5RSXlVVFxCdyxPT3EAOwMGFAWMADc0T0AvtMJ8/FDDNbYoriT83Pd3PP0TAJCIuIAcrLyympaugZGJsE2Gc6u7pdeN74BQQ2hSOEYXXifX6yVS6X+WQAkrgAKL4bQYdS1eohJrOKGw+GIwH9YH9HjLUbjKY7BLbfGLAnDVbrTak7YzIA===

For this case it is the same size as well, maybe I am missing something?

Reply:
> if the goal was to get the JIT to produce the equivalent of if `((utf8Text.Length | utf16Text.Length) == 0)`, why not write it that way?

Why write it the hard way when writing it the human-readable way produces identical optimized codegen? :)

> Can the JIT be improved to not be as sensitive to such variations?

You'd have to ask the JIT team, but per https://github.com/dotnet/runtime/issues/10950 it's large cost and expert skill level.

Reply:
@bartonjs, @jeffhandley, any opinions about the rule?  I could go either way.  On the one hand, I like the consistency, especially when looking at the variety of cases the rule found (e.g. span.Length < sizeof(byte)).  At the same time, much of that is "eye of the beholder", we've also now had to suppress several occurrences of the rule in this PR, and there's little perf benefit as Levi highlights.

Reply:
I agree with the "eye of the beholder" statement, and I think we've proven that this rule no longer benefits _performance_ across the board.  I'd like to make sure we classify the rule differently; I recommend "Maintainability" as the more appropriate category.

I also recommend we have the rule omit `Span<T>`, `ReadOnlySpan<T>`, `Memory<T>`, and `ReadOnlyMemory<T>` from the diagnostics, and we do so in a way that if we find we need to add to the list of excluded types, we can do so easily.  We will want to proactively watch for community feedback on this one too.  We can always dial back to the original set of types that we had identified where this rule was certainly better for performance.

Reply:
For `IsEmpty` vs `Count()` (a method/extension method) it's probably a performance win.  If we want to exclude the Span and Memory types as "IsEmpty and Count are the same, and we know it", that's perfectly reasonable to me.

Reply:
Thanks, I will update the analyzer to exclude those types. 

Is there any other type that we need to exclude? Or should we make a whitelist instead of a blacklist and just analyze certain types, e.g: only IEnumerable?

Reply:
Also, what does that decision means for this PR? Should I undo the changes to Span & Memory types?

Reply:
> Should I undo the changes to Span & Memory types?

@stephentoub, I think you preferred us to have uniform calls to these APIs throughout the runtime and libraries? If that's the case I'd keep the changes you made, even while special-casing _Span_ and _Memory_ not to trigger the analyzer.

Reply:
I personally have a small preference for it.  At the same time, it really is just a personal preference in general, and obviously some folks have the opposite opinion, and I don't want to force my preference here on others.  Given that, I'd suggest fixing the analyzer and then just redo'ing the PR with the new results, since I think 99.9% of the changes here are from span/memory (David, correct me if I'm wrong).